File: apps/api/internal/service/systemsettings.go
package service

import (
	"context"
	"errors"
	"time"

	"github.com/google/uuid"
	"github.com/lib/pq"

	"github.com/tolga/terp/internal/model"
)

var (
	ErrSystemSettingsNotFound      = errors.New("system settings not found")
	ErrInvalidBirthdayWindow       = errors.New("birthday window must be 0-90 days")
	ErrInvalidServerAliveTime      = errors.New("server alive expected completion time must be 0-1439 minutes")
	ErrInvalidServerAliveThreshold = errors.New("server alive threshold must be greater than 0")
	ErrInvalidDateRange            = errors.New("date_from must not be after date_to")
	ErrDateRangeTooLarge           = errors.New("date range must not exceed 366 days")
	ErrCleanupNoOrderIDs           = errors.New("at least one order_id is required")
)

// SystemSettingsLookup provides read-only access to system settings.
type SystemSettingsLookup interface {
	IsRoundingRelativeToPlan(ctx context.Context, tenantID uuid.UUID) (bool, error)
}

// systemSettingsRepository defines the interface for system settings data access.
type systemSettingsRepository interface {
	GetByTenantID(ctx context.Context, tenantID uuid.UUID) (*model.SystemSettings, error)
	Create(ctx context.Context, settings *model.SystemSettings) error
	Update(ctx context.Context, settings *model.SystemSettings) error
	GetOrCreate(ctx context.Context, tenantID uuid.UUID) (*model.SystemSettings, error)
}

// systemSettingsBookingRepo defines the booking repository interface for cleanup operations.
type systemSettingsBookingRepo interface {
	DeleteByDateRange(ctx context.Context, tenantID uuid.UUID, dateFrom, dateTo time.Time, employeeIDs []uuid.UUID) (int64, error)
	CountByDateRange(ctx context.Context, tenantID uuid.UUID, dateFrom, dateTo time.Time, employeeIDs []uuid.UUID) (int64, error)
}

// systemSettingsDailyValueRepo defines the daily value repository interface for cleanup operations.
type systemSettingsDailyValueRepo interface {
	DeleteByDateRange(ctx context.Context, tenantID uuid.UUID, dateFrom, dateTo time.Time, employeeIDs []uuid.UUID) (int64, error)
	CountByDateRange(ctx context.Context, tenantID uuid.UUID, dateFrom, dateTo time.Time, employeeIDs []uuid.UUID) (int64, error)
}

// systemSettingsEDPRepo defines the employee day plan repository interface for cleanup operations.
type systemSettingsEDPRepo interface {
	DeleteByDateRange(ctx context.Context, tenantID uuid.UUID, dateFrom, dateTo time.Time, employeeIDs []uuid.UUID) (int64, error)
}

// systemSettingsOrderRepo defines the order repository interface for cleanup operations.
type systemSettingsOrderRepo interface {
	BulkDelete(ctx context.Context, tenantID uuid.UUID, orderIDs []uuid.UUID) (int64, error)
	CountByIDs(ctx context.Context, tenantID uuid.UUID, orderIDs []uuid.UUID) (int64, error)
}

// systemSettingsRecalcService defines the recalc service interface for re-read operations.
type systemSettingsRecalcService interface {
	TriggerRecalcAll(ctx context.Context, tenantID uuid.UUID, from, to time.Time) (*RecalcResult, error)
	TriggerRecalcBatch(ctx context.Context, tenantID uuid.UUID, employeeIDs []uuid.UUID, from, to time.Time) *RecalcResult
}

// SystemSettingsService handles business logic for system settings and cleanup operations.
type SystemSettingsService struct {
	settingsRepo   systemSettingsRepository
	bookingRepo    systemSettingsBookingRepo
	dailyValueRepo systemSettingsDailyValueRepo
	edpRepo        systemSettingsEDPRepo
	orderRepo      systemSettingsOrderRepo
	recalcService  systemSettingsRecalcService
}

// NewSystemSettingsService creates a new SystemSettingsService.
func NewSystemSettingsService(
	settingsRepo systemSettingsRepository,
	bookingRepo systemSettingsBookingRepo,
	dailyValueRepo systemSettingsDailyValueRepo,
	edpRepo systemSettingsEDPRepo,
	orderRepo systemSettingsOrderRepo,
	recalcService systemSettingsRecalcService,
) *SystemSettingsService {
	return &SystemSettingsService{
		settingsRepo:   settingsRepo,
		bookingRepo:    bookingRepo,
		dailyValueRepo: dailyValueRepo,
		edpRepo:        edpRepo,
		orderRepo:      orderRepo,
		recalcService:  recalcService,
	}
}

// Get retrieves the system settings for a tenant, creating defaults if none exist.
func (s *SystemSettingsService) Get(ctx context.Context, tenantID uuid.UUID) (*model.SystemSettings, error) {
	return s.settingsRepo.GetOrCreate(ctx, tenantID)
}

// IsRoundingRelativeToPlan checks if rounding relative to plan is enabled for a tenant.
// Implements SystemSettingsLookup interface.
func (s *SystemSettingsService) IsRoundingRelativeToPlan(ctx context.Context, tenantID uuid.UUID) (bool, error) {
	settings, err := s.settingsRepo.GetOrCreate(ctx, tenantID)
	if err != nil {
		return false, err
	}
	return settings.RoundingRelativeToPlan, nil
}

// UpdateSystemSettingsInput represents the input for updating system settings.
type UpdateSystemSettingsInput struct {
	RoundingRelativeToPlan            *bool
	ErrorListEnabled                  *bool
	TrackedErrorCodes                 []string
	AutoFillOrderEndBookings          *bool
	BirthdayWindowDaysBefore          *int
	BirthdayWindowDaysAfter           *int
	FollowUpEntriesEnabled            *bool
	ProxyHost                         *string
	ProxyPort                         *int
	ProxyUsername                     *string
	ProxyPassword                     *string
	ProxyEnabled                      *bool
	ServerAliveEnabled                *bool
	ServerAliveExpectedCompletionTime *int
	ServerAliveThresholdMinutes       *int
	ServerAliveNotifyAdmins           *bool
}

// Update updates system settings with validation.
func (s *SystemSettingsService) Update(ctx context.Context, tenantID uuid.UUID, input UpdateSystemSettingsInput) (*model.SystemSettings, error) {
	settings, err := s.settingsRepo.GetOrCreate(ctx, tenantID)
	if err != nil {
		return nil, err
	}

	// Validate and apply changes
	if input.RoundingRelativeToPlan != nil {
		settings.RoundingRelativeToPlan = *input.RoundingRelativeToPlan
	}
	if input.ErrorListEnabled != nil {
		settings.ErrorListEnabled = *input.ErrorListEnabled
	}
	if input.TrackedErrorCodes != nil {
		settings.TrackedErrorCodes = pq.StringArray(input.TrackedErrorCodes)
	}
	if input.AutoFillOrderEndBookings != nil {
		settings.AutoFillOrderEndBookings = *input.AutoFillOrderEndBookings
	}
	if input.BirthdayWindowDaysBefore != nil {
		if *input.BirthdayWindowDaysBefore < 0 || *input.BirthdayWindowDaysBefore > 90 {
			return nil, ErrInvalidBirthdayWindow
		}
		settings.BirthdayWindowDaysBefore = *input.BirthdayWindowDaysBefore
	}
	if input.BirthdayWindowDaysAfter != nil {
		if *input.BirthdayWindowDaysAfter < 0 || *input.BirthdayWindowDaysAfter > 90 {
			return nil, ErrInvalidBirthdayWindow
		}
		settings.BirthdayWindowDaysAfter = *input.BirthdayWindowDaysAfter
	}
	if input.FollowUpEntriesEnabled != nil {
		settings.FollowUpEntriesEnabled = *input.FollowUpEntriesEnabled
	}
	if input.ProxyHost != nil {
		settings.ProxyHost = input.ProxyHost
	}
	if input.ProxyPort != nil {
		settings.ProxyPort = input.ProxyPort
	}
	if input.ProxyUsername != nil {
		settings.ProxyUsername = input.ProxyUsername
	}
	if input.ProxyPassword != nil {
		settings.ProxyPassword = input.ProxyPassword
	}
	if input.ProxyEnabled != nil {
		settings.ProxyEnabled = *input.ProxyEnabled
	}
	if input.ServerAliveEnabled != nil {
		settings.ServerAliveEnabled = *input.ServerAliveEnabled
	}
	if input.ServerAliveExpectedCompletionTime != nil {
		if *input.ServerAliveExpectedCompletionTime < 0 || *input.ServerAliveExpectedCompletionTime > 1439 {
			return nil, ErrInvalidServerAliveTime
		}
		settings.ServerAliveExpectedCompletionTime = input.ServerAliveExpectedCompletionTime
	}
	if input.ServerAliveThresholdMinutes != nil {
		if *input.ServerAliveThresholdMinutes <= 0 {
			return nil, ErrInvalidServerAliveThreshold
		}
		settings.ServerAliveThresholdMinutes = input.ServerAliveThresholdMinutes
	}
	if input.ServerAliveNotifyAdmins != nil {
		settings.ServerAliveNotifyAdmins = *input.ServerAliveNotifyAdmins
	}

	if err := s.settingsRepo.Update(ctx, settings); err != nil {
		return nil, err
	}
	return settings, nil
}

// CleanupDateRangeInput represents the input for date-range-based cleanup operations.
type CleanupDateRangeInput struct {
	DateFrom    time.Time
	DateTo      time.Time
	EmployeeIDs []uuid.UUID
	Confirm     bool // false = preview only
}

// CleanupOrdersInput represents the input for order-based cleanup operations.
type CleanupOrdersInput struct {
	OrderIDs []uuid.UUID
	Confirm  bool
}

// CleanupResult represents the outcome of a cleanup operation.
type CleanupResult struct {
	Operation     string
	AffectedCount int64
	Preview       bool
	Details       map[string]any
}

// validateDateRange validates cleanup date range constraints.
func validateDateRange(input CleanupDateRangeInput) error {
	if input.DateFrom.After(input.DateTo) {
		return ErrInvalidDateRange
	}
	daysDiff := input.DateTo.Sub(input.DateFrom).Hours() / 24
	if daysDiff > 366 {
		return ErrDateRangeTooLarge
	}
	return nil
}

// DeleteBookings deletes bookings in a date range.
func (s *SystemSettingsService) DeleteBookings(ctx context.Context, tenantID uuid.UUID, input CleanupDateRangeInput) (*CleanupResult, error) {
	if err := validateDateRange(input); err != nil {
		return nil, err
	}

	// Preview mode: just count
	if !input.Confirm {
		count, err := s.bookingRepo.CountByDateRange(ctx, tenantID, input.DateFrom, input.DateTo, input.EmployeeIDs)
		if err != nil {
			return nil, err
		}
		return &CleanupResult{
			Operation:     "delete_bookings",
			AffectedCount: count,
			Preview:       true,
		}, nil
	}

	// Execute
	count, err := s.bookingRepo.DeleteByDateRange(ctx, tenantID, input.DateFrom, input.DateTo, input.EmployeeIDs)
	if err != nil {
		return nil, err
	}
	return &CleanupResult{
		Operation:     "delete_bookings",
		AffectedCount: count,
		Preview:       false,
	}, nil
}

// DeleteBookingData deletes bookings, daily values, and employee day plans in a date range.
func (s *SystemSettingsService) DeleteBookingData(ctx context.Context, tenantID uuid.UUID, input CleanupDateRangeInput) (*CleanupResult, error) {
	if err := validateDateRange(input); err != nil {
		return nil, err
	}

	// Preview mode: count bookings (primary metric)
	if !input.Confirm {
		bookingCount, err := s.bookingRepo.CountByDateRange(ctx, tenantID, input.DateFrom, input.DateTo, input.EmployeeIDs)
		if err != nil {
			return nil, err
		}
		dvCount, err := s.dailyValueRepo.CountByDateRange(ctx, tenantID, input.DateFrom, input.DateTo, input.EmployeeIDs)
		if err != nil {
			return nil, err
		}
		return &CleanupResult{
			Operation:     "delete_booking_data",
			AffectedCount: bookingCount + dvCount,
			Preview:       true,
			Details: map[string]any{
				"bookings":     bookingCount,
				"daily_values": dvCount,
			},
		}, nil
	}

	// Execute all three deletions
	bookingCount, err := s.bookingRepo.DeleteByDateRange(ctx, tenantID, input.DateFrom, input.DateTo, input.EmployeeIDs)
	if err != nil {
		return nil, err
	}
	dvCount, err := s.dailyValueRepo.DeleteByDateRange(ctx, tenantID, input.DateFrom, input.DateTo, input.EmployeeIDs)
	if err != nil {
		return nil, err
	}
	edpCount, err := s.edpRepo.DeleteByDateRange(ctx, tenantID, input.DateFrom, input.DateTo, input.EmployeeIDs)
	if err != nil {
		return nil, err
	}

	return &CleanupResult{
		Operation:     "delete_booking_data",
		AffectedCount: bookingCount + dvCount + edpCount,
		Preview:       false,
		Details: map[string]any{
			"bookings":           bookingCount,
			"daily_values":       dvCount,
			"employee_day_plans": edpCount,
		},
	}, nil
}

// ReReadBookings re-triggers calculation for bookings in a date range.
func (s *SystemSettingsService) ReReadBookings(ctx context.Context, tenantID uuid.UUID, input CleanupDateRangeInput) (*CleanupResult, error) {
	if err := validateDateRange(input); err != nil {
		return nil, err
	}

	// Preview mode: count bookings that would be recalculated
	if !input.Confirm {
		count, err := s.bookingRepo.CountByDateRange(ctx, tenantID, input.DateFrom, input.DateTo, input.EmployeeIDs)
		if err != nil {
			return nil, err
		}
		return &CleanupResult{
			Operation:     "re_read_bookings",
			AffectedCount: count,
			Preview:       true,
		}, nil
	}

	// Execute recalculation
	var recalcResult *RecalcResult
	if len(input.EmployeeIDs) > 0 {
		recalcResult = s.recalcService.TriggerRecalcBatch(ctx, tenantID, input.EmployeeIDs, input.DateFrom, input.DateTo)
	} else {
		var err error
		recalcResult, err = s.recalcService.TriggerRecalcAll(ctx, tenantID, input.DateFrom, input.DateTo)
		if err != nil {
			return nil, err
		}
	}

	return &CleanupResult{
		Operation:     "re_read_bookings",
		AffectedCount: int64(recalcResult.ProcessedDays),
		Preview:       false,
		Details: map[string]any{
			"processed_days": recalcResult.ProcessedDays,
			"failed_days":    recalcResult.FailedDays,
		},
	}, nil
}

// MarkDeleteOrders deletes orders by IDs.
func (s *SystemSettingsService) MarkDeleteOrders(ctx context.Context, tenantID uuid.UUID, input CleanupOrdersInput) (*CleanupResult, error) {
	if len(input.OrderIDs) == 0 {
		return nil, ErrCleanupNoOrderIDs
	}

	// Preview mode: count matching orders
	if !input.Confirm {
		count, err := s.orderRepo.CountByIDs(ctx, tenantID, input.OrderIDs)
		if err != nil {
			return nil, err
		}
		return &CleanupResult{
			Operation:     "mark_delete_orders",
			AffectedCount: count,
			Preview:       true,
		}, nil
	}

	// Execute
	count, err := s.orderRepo.BulkDelete(ctx, tenantID, input.OrderIDs)
	if err != nil {
		return nil, err
	}
	return &CleanupResult{
		Operation:     "mark_delete_orders",
		AffectedCount: count,
		Preview:       false,
	}, nil
}

File: apps/web/src/server/routers/systemSettings.ts
/**
 * System Settings Router
 *
 * Provides singleton system settings get/update and cleanup operations via tRPC.
 *
 * Replaces the Go backend system settings endpoints:
 * - GET /system-settings -> systemSettings.get
 * - PUT /system-settings -> systemSettings.update
 * - POST /system-settings/cleanup/delete-bookings -> systemSettings.cleanupDeleteBookings
 * - POST /system-settings/cleanup/delete-booking-data -> systemSettings.cleanupDeleteBookingData
 * - POST /system-settings/cleanup/re-read-bookings -> systemSettings.cleanupReReadBookings
 * - POST /system-settings/cleanup/mark-delete-orders -> systemSettings.cleanupMarkDeleteOrders
 *
 * @see apps/api/internal/service/systemsettings.go
 */
import { z } from "zod"
import { TRPCError } from "@trpc/server"
import { createTRPCRouter, tenantProcedure } from "../trpc"
import { requirePermission } from "../middleware/authorization"
import { permissionIdByKey } from "../lib/permission-catalog"
import type { PrismaClient } from "@/generated/prisma/client"
import { RecalcService } from "../services/recalc"

// --- Permission Constants ---

const SETTINGS_MANAGE = permissionIdByKey("settings.manage")!

// --- Output Schemas ---

const systemSettingsOutputSchema = z.object({
  id: z.string().uuid(),
  tenantId: z.string().uuid(),
  roundingRelativeToPlan: z.boolean(),
  errorListEnabled: z.boolean(),
  trackedErrorCodes: z.array(z.string()),
  autoFillOrderEndBookings: z.boolean(),
  birthdayWindowDaysBefore: z.number(),
  birthdayWindowDaysAfter: z.number(),
  followUpEntriesEnabled: z.boolean(),
  proxyHost: z.string().nullable(),
  proxyPort: z.number().nullable(),
  proxyUsername: z.string().nullable(),
  proxyEnabled: z.boolean(),
  serverAliveEnabled: z.boolean(),
  serverAliveExpectedCompletionTime: z.number().nullable(),
  serverAliveThresholdMinutes: z.number().nullable(),
  serverAliveNotifyAdmins: z.boolean(),
  createdAt: z.date(),
  updatedAt: z.date(),
})

const cleanupResultSchema = z.object({
  operation: z.string(),
  affectedCount: z.number(),
  preview: z.boolean(),
  details: z.record(z.string(), z.unknown()).optional(),
})

// --- Input Schemas ---

const updateSettingsInputSchema = z.object({
  roundingRelativeToPlan: z.boolean().optional(),
  errorListEnabled: z.boolean().optional(),
  trackedErrorCodes: z.array(z.string()).optional(),
  autoFillOrderEndBookings: z.boolean().optional(),
  birthdayWindowDaysBefore: z.number().int().min(0).max(90).optional(),
  birthdayWindowDaysAfter: z.number().int().min(0).max(90).optional(),
  followUpEntriesEnabled: z.boolean().optional(),
  proxyHost: z.string().max(255).nullable().optional(),
  proxyPort: z.number().int().nullable().optional(),
  proxyUsername: z.string().max(255).nullable().optional(),
  proxyPassword: z.string().max(255).nullable().optional(),
  proxyEnabled: z.boolean().optional(),
  serverAliveEnabled: z.boolean().optional(),
  serverAliveExpectedCompletionTime: z
    .number()
    .int()
    .min(0)
    .max(1439)
    .nullable()
    .optional(),
  serverAliveThresholdMinutes: z
    .number()
    .int()
    .min(1)
    .nullable()
    .optional(),
  serverAliveNotifyAdmins: z.boolean().optional(),
})

const cleanupDateRangeInputSchema = z.object({
  dateFrom: z.string().date(),
  dateTo: z.string().date(),
  employeeIds: z.array(z.string().uuid()).optional(),
  confirm: z.boolean(),
})

const cleanupOrdersInputSchema = z.object({
  orderIds: z.array(z.string().uuid()).min(1, "At least one order ID required"),
  confirm: z.boolean(),
})

// --- Helpers ---

/**
 * Singleton getOrCreate pattern for system settings.
 * Returns existing settings for the tenant, or creates defaults.
 */
async function getOrCreateSettings(
  prisma: PrismaClient,
  tenantId: string
) {
  const existing = await prisma.systemSetting.findUnique({
    where: { tenantId },
  })
  if (existing) return existing

  return prisma.systemSetting.create({
    data: { tenantId },
  })
}

/**
 * Maps a SystemSetting record to the output shape, omitting proxyPassword.
 */
function mapToOutput(s: Record<string, unknown>) {
  return {
    id: s.id as string,
    tenantId: s.tenantId as string,
    roundingRelativeToPlan: s.roundingRelativeToPlan as boolean,
    errorListEnabled: s.errorListEnabled as boolean,
    trackedErrorCodes: s.trackedErrorCodes as string[],
    autoFillOrderEndBookings: s.autoFillOrderEndBookings as boolean,
    birthdayWindowDaysBefore: s.birthdayWindowDaysBefore as number,
    birthdayWindowDaysAfter: s.birthdayWindowDaysAfter as number,
    followUpEntriesEnabled: s.followUpEntriesEnabled as boolean,
    proxyHost: (s.proxyHost as string | null) ?? null,
    proxyPort: (s.proxyPort as number | null) ?? null,
    proxyUsername: (s.proxyUsername as string | null) ?? null,
    // proxyPassword intentionally omitted (matches Go json:"-" behavior)
    proxyEnabled: s.proxyEnabled as boolean,
    serverAliveEnabled: s.serverAliveEnabled as boolean,
    serverAliveExpectedCompletionTime:
      (s.serverAliveExpectedCompletionTime as number | null) ?? null,
    serverAliveThresholdMinutes:
      (s.serverAliveThresholdMinutes as number | null) ?? null,
    serverAliveNotifyAdmins: s.serverAliveNotifyAdmins as boolean,
    createdAt: s.createdAt as Date,
    updatedAt: s.updatedAt as Date,
  }
}

/**
 * Validates a cleanup date range: dateFrom <= dateTo, range <= 366 days.
 */
function validateDateRange(dateFrom: string, dateTo: string) {
  const from = new Date(dateFrom)
  const to = new Date(dateTo)

  if (from > to) {
    throw new TRPCError({
      code: "BAD_REQUEST",
      message: "dateFrom must be before or equal to dateTo",
    })
  }

  const diffMs = to.getTime() - from.getTime()
  const diffDays = diffMs / (1000 * 60 * 60 * 24)
  if (diffDays > 366) {
    throw new TRPCError({
      code: "BAD_REQUEST",
      message: "Date range must not exceed 366 days",
    })
  }
}

/**
 * Deletes bookings for a tenant within a date range, with optional employee filter.
 */
async function deleteBookings(
  prisma: PrismaClient,
  tenantId: string,
  dateFrom: string,
  dateTo: string,
  employeeIds?: string[]
): Promise<number> {
  const where: Record<string, unknown> = {
    tenantId,
    bookingDate: {
      gte: new Date(dateFrom),
      lte: new Date(dateTo),
    },
  }
  if (employeeIds && employeeIds.length > 0) {
    where.employeeId = { in: employeeIds }
  }
  const result = await prisma.booking.deleteMany({ where })
  return result.count
}

/**
 * Counts bookings for a tenant within a date range, with optional employee filter.
 */
async function countBookings(
  prisma: PrismaClient,
  tenantId: string,
  dateFrom: string,
  dateTo: string,
  employeeIds?: string[]
): Promise<number> {
  const where: Record<string, unknown> = {
    tenantId,
    bookingDate: {
      gte: new Date(dateFrom),
      lte: new Date(dateTo),
    },
  }
  if (employeeIds && employeeIds.length > 0) {
    where.employeeId = { in: employeeIds }
  }
  return prisma.booking.count({ where })
}

/**
 * Deletes daily_values for a tenant within a date range, with optional employee filter.
 * Fixes pre-existing bug: raw SQL referenced column 'date' but actual column is 'value_date'.
 */
async function deleteDailyValues(
  prisma: PrismaClient,
  tenantId: string,
  dateFrom: string,
  dateTo: string,
  employeeIds?: string[]
): Promise<number> {
  const where: Record<string, unknown> = {
    tenantId,
    valueDate: {
      gte: new Date(dateFrom),
      lte: new Date(dateTo),
    },
  }
  if (employeeIds && employeeIds.length > 0) {
    where.employeeId = { in: employeeIds }
  }
  const result = await prisma.dailyValue.deleteMany({ where })
  return result.count
}

/**
 * Counts daily_values for a tenant within a date range, with optional employee filter.
 * Fixes pre-existing bug: raw SQL referenced column 'date' but actual column is 'value_date'.
 */
async function countDailyValues(
  prisma: PrismaClient,
  tenantId: string,
  dateFrom: string,
  dateTo: string,
  employeeIds?: string[]
): Promise<number> {
  const where: Record<string, unknown> = {
    tenantId,
    valueDate: {
      gte: new Date(dateFrom),
      lte: new Date(dateTo),
    },
  }
  if (employeeIds && employeeIds.length > 0) {
    where.employeeId = { in: employeeIds }
  }
  return prisma.dailyValue.count({ where })
}

/**
 * Deletes employee_day_plans for a tenant within a date range, with optional employee filter.
 * Uses Prisma model instead of raw SQL.
 */
async function deleteEmployeeDayPlans(
  prisma: PrismaClient,
  tenantId: string,
  dateFrom: string,
  dateTo: string,
  employeeIds?: string[]
): Promise<number> {
  const where: Record<string, unknown> = {
    tenantId,
    planDate: {
      gte: new Date(dateFrom),
      lte: new Date(dateTo),
    },
  }

  if (employeeIds && employeeIds.length > 0) {
    where.employeeId = { in: employeeIds }
  }

  const result = await prisma.employeeDayPlan.deleteMany({ where })
  return result.count
}

// --- Router ---

export const systemSettingsRouter = createTRPCRouter({
  /**
   * systemSettings.get -- Returns the singleton system settings for the tenant.
   *
   * Creates default settings on first access (getOrCreate pattern).
   * ProxyPassword is never returned.
   *
   * Requires: settings.manage permission
   */
  get: tenantProcedure
    .use(requirePermission(SETTINGS_MANAGE))
    .output(systemSettingsOutputSchema)
    .query(async ({ ctx }) => {
      const tenantId = ctx.tenantId!
      const settings = await getOrCreateSettings(ctx.prisma, tenantId)
      return mapToOutput(settings as unknown as Record<string, unknown>)
    }),

  /**
   * systemSettings.update -- Partially updates system settings.
   *
   * Uses getOrCreate to ensure the singleton exists before updating.
   * ProxyPassword is never returned.
   *
   * Requires: settings.manage permission
   */
  update: tenantProcedure
    .use(requirePermission(SETTINGS_MANAGE))
    .input(updateSettingsInputSchema)
    .output(systemSettingsOutputSchema)
    .mutation(async ({ ctx, input }) => {
      const tenantId = ctx.tenantId!

      // Ensure settings exist
      const existing = await getOrCreateSettings(ctx.prisma, tenantId)

      // Build partial update data
      const data: Record<string, unknown> = {}

      if (input.roundingRelativeToPlan !== undefined) {
        data.roundingRelativeToPlan = input.roundingRelativeToPlan
      }
      if (input.errorListEnabled !== undefined) {
        data.errorListEnabled = input.errorListEnabled
      }
      if (input.trackedErrorCodes !== undefined) {
        data.trackedErrorCodes = input.trackedErrorCodes
      }
      if (input.autoFillOrderEndBookings !== undefined) {
        data.autoFillOrderEndBookings = input.autoFillOrderEndBookings
      }
      if (input.birthdayWindowDaysBefore !== undefined) {
        data.birthdayWindowDaysBefore = input.birthdayWindowDaysBefore
      }
      if (input.birthdayWindowDaysAfter !== undefined) {
        data.birthdayWindowDaysAfter = input.birthdayWindowDaysAfter
      }
      if (input.followUpEntriesEnabled !== undefined) {
        data.followUpEntriesEnabled = input.followUpEntriesEnabled
      }
      if (input.proxyHost !== undefined) {
        data.proxyHost = input.proxyHost
      }
      if (input.proxyPort !== undefined) {
        data.proxyPort = input.proxyPort
      }
      if (input.proxyUsername !== undefined) {
        data.proxyUsername = input.proxyUsername
      }
      if (input.proxyPassword !== undefined) {
        data.proxyPassword = input.proxyPassword
      }
      if (input.proxyEnabled !== undefined) {
        data.proxyEnabled = input.proxyEnabled
      }
      if (input.serverAliveEnabled !== undefined) {
        data.serverAliveEnabled = input.serverAliveEnabled
      }
      if (input.serverAliveExpectedCompletionTime !== undefined) {
        data.serverAliveExpectedCompletionTime =
          input.serverAliveExpectedCompletionTime
      }
      if (input.serverAliveThresholdMinutes !== undefined) {
        data.serverAliveThresholdMinutes = input.serverAliveThresholdMinutes
      }
      if (input.serverAliveNotifyAdmins !== undefined) {
        data.serverAliveNotifyAdmins = input.serverAliveNotifyAdmins
      }

      const updated = await ctx.prisma.systemSetting.update({
        where: { id: existing.id },
        data,
      })

      return mapToOutput(updated as unknown as Record<string, unknown>)
    }),

  /**
   * systemSettings.cleanupDeleteBookings -- Delete bookings in a date range.
   *
   * Preview mode (confirm: false): returns count of affected bookings.
   * Execute mode (confirm: true): deletes bookings and returns count.
   *
   * Requires: settings.manage permission
   */
  cleanupDeleteBookings: tenantProcedure
    .use(requirePermission(SETTINGS_MANAGE))
    .input(cleanupDateRangeInputSchema)
    .output(cleanupResultSchema)
    .mutation(async ({ ctx, input }) => {
      const tenantId = ctx.tenantId!
      validateDateRange(input.dateFrom, input.dateTo)

      if (!input.confirm) {
        // Preview mode: count bookings
        const count = await countBookings(
          ctx.prisma,
          tenantId,
          input.dateFrom,
          input.dateTo,
          input.employeeIds
        )

        return {
          operation: "delete_bookings",
          affectedCount: count,
          preview: true,
        }
      }

      // Execute mode: delete bookings
      const deleted = await deleteBookings(
        ctx.prisma,
        tenantId,
        input.dateFrom,
        input.dateTo,
        input.employeeIds
      )

      return {
        operation: "delete_bookings",
        affectedCount: deleted,
        preview: false,
      }
    }),

  /**
   * systemSettings.cleanupDeleteBookingData -- Delete bookings, daily values,
   * and employee day plans in a date range.
   *
   * Preview mode: returns counts of bookings and daily values.
   * Execute mode: deletes all three entity types.
   *
   * Requires: settings.manage permission
   */
  cleanupDeleteBookingData: tenantProcedure
    .use(requirePermission(SETTINGS_MANAGE))
    .input(cleanupDateRangeInputSchema)
    .output(cleanupResultSchema)
    .mutation(async ({ ctx, input }) => {
      const tenantId = ctx.tenantId!
      validateDateRange(input.dateFrom, input.dateTo)

      if (!input.confirm) {
        // Preview mode: count bookings + daily values
        const [bookingsCount, dailyValuesCount] = await Promise.all([
          countBookings(
            ctx.prisma,
            tenantId,
            input.dateFrom,
            input.dateTo,
            input.employeeIds
          ),
          countDailyValues(
            ctx.prisma,
            tenantId,
            input.dateFrom,
            input.dateTo,
            input.employeeIds
          ),
        ])

        return {
          operation: "delete_booking_data",
          affectedCount: bookingsCount + dailyValuesCount,
          preview: true,
          details: {
            bookings: bookingsCount,
            dailyValues: dailyValuesCount,
          },
        }
      }

      // Execute mode: delete bookings + daily values + employee day plans
      const [deletedBookings, deletedDailyValues, deletedEdps] =
        await Promise.all([
          deleteBookings(
            ctx.prisma,
            tenantId,
            input.dateFrom,
            input.dateTo,
            input.employeeIds
          ),
          deleteDailyValues(
            ctx.prisma,
            tenantId,
            input.dateFrom,
            input.dateTo,
            input.employeeIds
          ),
          deleteEmployeeDayPlans(
            ctx.prisma,
            tenantId,
            input.dateFrom,
            input.dateTo,
            input.employeeIds
          ),
        ])

      return {
        operation: "delete_booking_data",
        affectedCount: deletedBookings + deletedDailyValues + deletedEdps,
        preview: false,
        details: {
          bookings: deletedBookings,
          dailyValues: deletedDailyValues,
          employeeDayPlans: deletedEdps,
        },
      }
    }),

  /**
   * systemSettings.cleanupReReadBookings -- Re-read bookings in a date range.
   *
   * Preview mode: returns count of bookings.
   * Execute mode: NOT_IMPLEMENTED (recalculation service not yet ported).
   *
   * Requires: settings.manage permission
   */
  cleanupReReadBookings: tenantProcedure
    .use(requirePermission(SETTINGS_MANAGE))
    .input(cleanupDateRangeInputSchema)
    .output(cleanupResultSchema)
    .mutation(async ({ ctx, input }) => {
      const tenantId = ctx.tenantId!
      validateDateRange(input.dateFrom, input.dateTo)

      if (!input.confirm) {
        // Preview mode: count bookings
        const count = await countBookings(
          ctx.prisma,
          tenantId,
          input.dateFrom,
          input.dateTo,
          input.employeeIds
        )

        return {
          operation: "re_read_bookings",
          affectedCount: count,
          preview: true,
        }
      }

      // Execute mode: recalculate bookings
      // @see ZMI-TICKET-243
      const recalcService = new RecalcService(ctx.prisma)

      const fromDate = new Date(input.dateFrom)
      const toDate = new Date(input.dateTo)

      let result
      if (input.employeeIds && input.employeeIds.length > 0) {
        result = await recalcService.triggerRecalcBatch(
          tenantId,
          input.employeeIds,
          fromDate,
          toDate,
        )
      } else {
        result = await recalcService.triggerRecalcAll(
          tenantId,
          fromDate,
          toDate,
        )
      }

      return {
        operation: "re_read_bookings",
        affectedCount: result.processedDays,
        preview: false,
      }
    }),

  /**
   * systemSettings.cleanupMarkDeleteOrders -- Delete orders by IDs.
   *
   * Preview mode: returns count of matching orders.
   * Execute mode: deletes the orders.
   * Uses Prisma model since Order has a Prisma model.
   *
   * Requires: settings.manage permission
   */
  cleanupMarkDeleteOrders: tenantProcedure
    .use(requirePermission(SETTINGS_MANAGE))
    .input(cleanupOrdersInputSchema)
    .output(cleanupResultSchema)
    .mutation(async ({ ctx, input }) => {
      const tenantId = ctx.tenantId!

      if (!input.confirm) {
        // Preview mode: count orders
        const count = await ctx.prisma.order.count({
          where: {
            id: { in: input.orderIds },
            tenantId,
          },
        })

        return {
          operation: "mark_delete_orders",
          affectedCount: count,
          preview: true,
        }
      }

      // Execute mode: delete orders
      const result = await ctx.prisma.order.deleteMany({
        where: {
          id: { in: input.orderIds },
          tenantId,
        },
      })

      return {
        operation: "mark_delete_orders",
        affectedCount: result.count,
        preview: false,
      }
    }),
})
